File: doc-examples/mongo-example-java/src/test/java/example/PersonRepositorySpec.java
package example;

import io.micronaut.data.repository.jpa.criteria.PredicateSpecification;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

import static example.PersonRepository.Specifications.ageIsLessThan;
import static example.PersonRepository.Specifications.interestsContains;
import static example.PersonRepository.Specifications.nameEquals;
import static example.PersonRepository.Specifications.setNewName;
import static io.micronaut.data.repository.jpa.criteria.PredicateSpecification.not;
import static io.micronaut.data.repository.jpa.criteria.PredicateSpecification.where;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

@MicronautTest
class PersonRepositorySpec {

    @Inject
    PersonRepository personRepository;

    @BeforeEach
    void beforeEach() {
        personRepository.saveAll(Arrays.asList(
                new Person(
                        "Denis",
                        13
                ),
                new Person(
                        "Josh",
                        22
                ).withInterests(Arrays.asList("sports", "music", "hiking"))
        ));
    }

    @AfterEach
    void afterEach() {
        personRepository.deleteAll();
    }

    @Test
    void testFind() {
        // tag::find[]
        Person denis = personRepository.findOne(nameEquals("Denis")).orElse(null);

        long countAgeLess30 = personRepository.count(ageIsLessThan(30));

        long countAgeLess20 = personRepository.count(ageIsLessThan(20));

        long countAgeLess30NotDenis = personRepository.count(ageIsLessThan(30).and(not(nameEquals("Denis"))));

        List<Person> people = personRepository.findAll(where(nameEquals("Denis").or(nameEquals("Josh"))));
        // end::find[]

        assertNotNull(denis);
        assertEquals(2, countAgeLess30);
        assertEquals(1, countAgeLess20);
        assertEquals(1, countAgeLess30NotDenis);
        assertEquals(2, people.size());
    }

    @Test
    void testDelete() {
        List<Person> all = personRepository.findAll((PredicateSpecification<Person>) null);
        assertEquals(2, all.size());

        // tag::delete[]
        long recordsDeleted = personRepository.deleteAll(where(nameEquals("Denis")));
        // end::delete[]

        assertEquals(1, recordsDeleted);

        all = personRepository.findAll((PredicateSpecification<Person>) null);
        assertEquals(1, all.size());
    }

    @Test
    void testUpdate() {
        List<Person> all = personRepository.findAll((PredicateSpecification<Person>) null);
        assertEquals(2, all.size());
        assertTrue(all.stream().anyMatch(p -> p.getName().equals("Denis")));
        assertTrue(all.stream().anyMatch(p -> p.getName().equals("Josh")));

        // tag::update[]
        long recordsUpdated = personRepository.updateAll(setNewName("Steven").where(nameEquals("Denis")));
        // end::update[]

        assertEquals(1, recordsUpdated);

        all = personRepository.findAll((PredicateSpecification<Person>) null);
        assertEquals(2, all.size());
        assertTrue(all.stream().anyMatch(p -> p.getName().equals("Steven")));
        assertTrue(all.stream().anyMatch(p -> p.getName().equals("Josh")));
    }

    @Test
    void testArrayContains() {
        List<Person> people = personRepository.findByInterestsCollectionContains("sports");
        assertEquals(1, people.size());
        assertEquals("Josh", people.get(0).getName());

        people = personRepository.findByInterestsCollectionContains("flying");
        assertTrue(people.isEmpty());

        // Using specification
        people = personRepository.findAll(interestsContains("hiking"));
        assertEquals(1, people.size());
        assertEquals("Josh", people.get(0).getName());
    }

}

File: doc-examples/mongo-example-kotlin/src/test/kotlin/example/PersonRepositorySpec.kt
package example

import example.PersonRepository.Specifications.ageIsLessThan
import example.PersonRepository.Specifications.interestsContains
import example.PersonRepository.Specifications.nameEquals
import example.PersonRepository.Specifications.updateName
import io.micronaut.data.repository.jpa.criteria.PredicateSpecification
import io.micronaut.data.repository.jpa.criteria.PredicateSpecification.not
import io.micronaut.data.runtime.criteria.get
import io.micronaut.data.runtime.criteria.query
import io.micronaut.data.runtime.criteria.where
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import jakarta.inject.Inject
import org.junit.jupiter.api.*
import java.util.*

@MicronautTest
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class PersonRepositorySpec : AbstractMongoSpec() {
    @Inject
    private lateinit var personRepository: PersonRepository

    @BeforeEach
    fun beforeEach() {
        personRepository.saveAll(listOf(
                Person(
                        "Denis",
                        13
                ),
                Person(
                    null,
                    "Josh",
                    22,
                    listOf("music", "sports", "hiking")
                )
        ))
    }

    @AfterEach
    fun afterEach() {
        personRepository.deleteAll()
    }

    @Test
    fun testFind() {
        // tag::find[]
        val denis: Person? = personRepository.findOne(nameEquals("Denis")).orElse(null)

        val countAgeLess30: Long = personRepository.count(ageIsLessThan(30))

        val countAgeLess20: Long = personRepository.count(ageIsLessThan(20))

        val countAgeLess30NotDenis: Long = personRepository.count(ageIsLessThan(30).and(not(nameEquals("Denis"))))

        val people = personRepository.findAll(PredicateSpecification.where(nameEquals("Denis").or(nameEquals("Josh"))))
        // end::find[]
        Assertions.assertNotNull(denis)
        Assertions.assertEquals(2, countAgeLess30)
        Assertions.assertEquals(1, countAgeLess20)
        Assertions.assertEquals(1, countAgeLess30NotDenis)
        Assertions.assertEquals(2, people.size)
    }

    @Test
    fun testFindDto() {
        val stats = personRepository.findOne(query<Person, PersonAgeStatsDto> {
            multiselect(
                    max(Person::age).alias(PersonAgeStatsDto::maxAge),
                    min(Person::age).alias(PersonAgeStatsDto::minAge),
                    avg(Person::age).alias(PersonAgeStatsDto::avgAge)
            )
            where {
                or {
                    root[Person::name] eq "Denis"
                    root[Person::name] eq "Josh"
                }
            }
        })

        Assertions.assertEquals(22, stats.maxAge)
        Assertions.assertEquals(13, stats.minAge)
        Assertions.assertEquals(17.5, stats.avgAge)
    }

    @Test
    fun testDelete() {
        val empty: PredicateSpecification<Person>? = null
        var all = personRepository.findAll(empty)
        Assertions.assertEquals(2, all.size)

        // tag::delete[]
        val recordsDeleted = personRepository.deleteAll(PredicateSpecification.where(nameEquals("Denis")))
        // end::delete[]
        Assertions.assertEquals(1, recordsDeleted)
        all = personRepository.findAll(empty)
        Assertions.assertEquals(1, all.size)
    }

    @Test
    fun testUpdate() {
        val empty: PredicateSpecification<Person>? = null
        var all = personRepository.findAll(empty)
        Assertions.assertEquals(2, all.size)
        Assertions.assertTrue(all.stream().anyMatch { p: Person -> p.name == "Denis" })
        Assertions.assertTrue(all.stream().anyMatch { p: Person -> p.name == "Josh" })

        // tag::update[]
        val recordsUpdated = personRepository.updateAll(updateName("Steven", "Denis"))
        // end::update[]
        Assertions.assertEquals(1, recordsUpdated)
        all = personRepository.findAll(empty)
        Assertions.assertEquals(2, all.size)
        Assertions.assertTrue(all.stream().anyMatch { p: Person -> p.name == "Steven" })
        Assertions.assertTrue(all.stream().anyMatch { p: Person -> p.name == "Josh" })
    }

    @Test
    fun testDeleteUsingCriteriaBuilder() {
        val empty: PredicateSpecification<Person>? = null
        var all = personRepository.findAll(empty)
        Assertions.assertEquals(2, all.size)

        // tag::delete[]
        val recordsDeleted = personRepository.deleteAll(where {
            root[Person::name] eq "Denis"
        })
        // end::delete[]
        Assertions.assertEquals(1, recordsDeleted)
        all = personRepository.findAll(empty)
        Assertions.assertEquals(1, all.size)
    }

    @Test
    fun testDeleteUsingCriteriaBuilder2() {
        val empty: PredicateSpecification<Person>? = null
        var all = personRepository.findAll(empty)
        Assertions.assertEquals(2, all.size)

        // tag::delete[]
        val recordsDeleted = personRepository.deleteAll(where {
            root[Person::name] eq "Denis"
        })
        // end::delete[]
        Assertions.assertEquals(1, recordsDeleted)
        all = personRepository.findAll(empty)
        Assertions.assertEquals(1, all.size)
    }

    @Test
    fun testArrayContains() {
        var people = personRepository.findByInterestsCollectionContains("sports")
        Assertions.assertEquals(1, people.size)
        Assertions.assertEquals("Josh", people[0].name)

        people = personRepository.findByInterestsCollectionContains("flying")
        Assertions.assertTrue(people.isEmpty())

        // Using specification
        people = personRepository.findAll(interestsContains( "hiking"))
        Assertions.assertEquals(1, people.size)
        Assertions.assertEquals("Josh", people[0].name)
    }
}
